xarray/backends/api.py

```diff
@@ -15,35 +15,53 @@
 from ..core.combine import auto_combine
 from ..core.utils import close_on_error, is_remote_uri
 from ..core.pycompat import basestring, path_type
+from ..core.options import OPTIONS
 
 DATAARRAY_NAME = '__xarray_dataarray_name__'
 DATAARRAY_VARIABLE = '__xarray_dataarray_variable__'
 
 
+def _guess_file_format(path):
+    """Make an educated guess about the file format based on the first few
+    bytes.
+
+    If the file does not exist, assume netcdf3.
+    """
+    try:
+        f = open(path, 'rb')
+    except IOError:
+        return 'netcdf3'
+
+    with f:
+        # Ten bytes should be enough...
+        sig = f.read(10)
+
+    if sig[:8] == b'\x89HDF\r\n\x1a\n':
+        return 'hdf5'
+    elif sig[:4] == b'CDF\x02' or sig[:4] == b'CDF\x01':
+        return 'netcdf3'
+    else:
+        return 'unknown'
+
+
 def _get_default_engine(path, allow_remote=False):
     if allow_remote and is_remote_uri(path):  # pragma: no cover
-        try:
-            import netCDF4
-            engine = 'netcdf4'
-        except ImportError:
-            try:
-                import pydap
-                engine = 'pydap'
-            except ImportError:
-                raise ValueError('netCDF4 or pydap is required for accessing '
-                                 'remote datasets via OPeNDAP')
+        for engine in OPTIONS['io_engines']:
+            if engine in ('netcdf4', 'pydap'):
+                return engine
+        else:
+            raise ValueError('netCDF4 or pydap is required for accessing '
+                             'remote datasets via OPeNDAP')
     else:
-        try:
-            import netCDF4
-            engine = 'netcdf4'
-        except ImportError:  # pragma: no cover
-            try:
-                import scipy.io.netcdf
-                engine = 'scipy'
-            except ImportError:
-                raise ValueError('cannot read or write netCDF files without '
-                                 'netCDF4-python or scipy installed')
-    return engine
+        filefmt = _guess_file_format(path)
+        for engine in OPTIONS['io_engines']:
+            if filefmt == 'hdf5' and engine in ('netcdf4', 'h5netcdf'):
+                return engine
+            elif filefmt == 'netcdf3' and engine in ('netcdf4', 'scipy'):
+                return engine
+        else:
+            raise ValueError('cannot read or write netCDF files without '
+                             'netCDF4-python, scipy, or h5netcdf installed')
```

Comment:
Also, please include the list of available default IO engines in the error message.